concordium-consensus/src/Concordium/GlobalState/Persistent/ReleaseSchedule.hs

```diff
@@ -0,0 +1,415 @@
+
+-- |Migrate a release schedule from one protocol version to another, given by a
+-- 'ReleaseScheduleMigration'.
+migrateReleaseSchedule ::
+    (SupportMigration m t) =>
+    ReleaseScheduleMigration m oldpv pv ->
+    ReleaseSchedule oldpv ->
+    t m (ReleaseSchedule pv)
+migrateReleaseSchedule RSMLegacyToLegacy (ReleaseScheduleP0 rs) =
+    ReleaseScheduleP0 <$!> migrateReference return rs
+migrateReleaseSchedule (RSMLegacyToNew resolveAcc) (ReleaseScheduleP0 rsRef) = do
+    rs <- lift $ refLoad rsRef
+    rsMapList <- lift $ forM (Map.toList (rsMap rs)) $ \(ts, addrs) -> do
+        aiList <- mapM resolveAcc (Set.toList addrs)
+        return (ts, AccountSet (Set.fromList aiList))
+    newMap <- Trie.fromList rsMapList
+    return $!
+        ReleaseScheduleP5
+            NewReleaseSchedule
+                { rs5FirstTimestamp = rsFirstTimestamp rs,
+                  rs5Map = newMap
+                }
+migrateReleaseSchedule RSMNewToNew (ReleaseScheduleP5 rs) = do
+    newMap <- Trie.migrateTrieN True return (rs5Map rs)
+    return $!
+        ReleaseScheduleP5
+            NewReleaseSchedule
+                { rs5FirstTimestamp = rs5FirstTimestamp rs,
+                  rs5Map = newMap
+                }
+
+-- |Make a persistent release schedule from an in-memory release schedule.
+-- This takes a function for resolving account indexes to account addresses, which is used for
+-- protocol versions that use the legacy release schedule.
+makePersistentReleaseSchedule ::
+    forall m pv.
+    (IsProtocolVersion pv, MonadBlobStore m) =>
+    -- |Function to resolve an account address from an account index
+    (AccountIndex -> AccountAddress) ->
+    -- |In-memory release schedule
+    Transient.ReleaseSchedule ->
+    m (ReleaseSchedule pv)
+makePersistentReleaseSchedule getAddr tRS = case protocolVersion @pv of
+    SP1 -> rsP0
+    SP2 -> rsP0
+    SP3 -> rsP0
+    SP4 -> rsP0
+    SP5 -> rsP1
+  where
+    rsP0 :: (RSAccountRef pv ~ AccountAddress) => m (ReleaseSchedule pv)
+    rsP0 = do
+        let tf !entries accIds =
+                let newSet = Set.map getAddr accIds
+                 in (entries + fromIntegral (Set.size newSet), newSet)
+            (rsEntryCount, rsMap) = Map.mapAccum tf 0 (Transient.rsMap tRS)
+        rsRef <-
+            refMake $!
+                LegacyReleaseSchedule
+                    { rsFirstTimestamp = Transient.rsFirstTimestamp tRS,
+                      ..
+                    }
+        return $! ReleaseScheduleP0 rsRef
+    rsP1 :: (RSAccountRef pv ~ AccountIndex) => m (ReleaseSchedule pv)
+    rsP1 = do
+        rs5Map <- Trie.fromList (second AccountSet <$> Map.toList (Transient.rsMap tRS))
+        return $!
+            ReleaseScheduleP5
+                NewReleaseSchedule
+                    { rs5FirstTimestamp = Transient.rsFirstTimestamp tRS,
+                      ..
+                    }
+
+-- |(For testing purposes) get the map of the earliest scheduled releases of each account.
```

Comment:
Maybe add a warning pragma like we do for similar functions.

Reply:
I don't think it's warranted in this case. "For testing purposes" is mainly meant as an explanation of why this code exists in the first place.